packages/lib-classifier/src/store/TutorialStore.js

```diff
@@ -142,7 +142,8 @@ const TutorialStore = types
         const response = yield tutorialsClient.get({ workflowId: workflow.id })
         const { tutorials } = response.body
         if (tutorials && tutorials.length > 0) {
-          yield Promise.all(tutorials.map(tutorial => self.fetchMedia(tutorial)))
+          const mediaRequests = tutorials.map(tutorial => flow(fetchMedia)(tutorial))
+          yield Promise.all(mediaRequests)
```

Comment:
Test expectations are being evaluated here, before the tutorial is set and `showTutorialInModal` is called. The workaround seems to be to stub `fetchMedia` to return `Promise.resolve()` and I can't see why that's needed to pass the tests.

Reply:
via https://github.com/mobxjs/mobx-state-tree/blob/master/docs/async-actions.md#using-generators

> `Yield` always needs to return a `Promise`.

...but there are several states in which `fetchMedia` doesn't (e.g. no tutorial), unless it's always stubbed to return a resolved promise 💡

Reply:
But `tutorials.getAttachedImages` is stubbed to return a resolved promise in the tests. That's what's confusing me. https://github.com/zooniverse/front-end-monorepo/blob/ebd450a09dd5540ff62f8771d1b5409b70acba1a/packages/lib-classifier/src/store/TutorialStore.spec.js#L73-L79

Reply:
I'm seeing that resolve with the mock media, then the expectation being evaluated, then `setTutorial` being called.

Reply:
Hmmm. I've just tried running `it('should request for the media if there are tutorials')` in isolation with `.only`, and it passed; could the test state be dirty?

Reply:
That's not one of the failing tests.

Try this one:
```
Model > TutorialStore
       should set the tutorial if there is a workflow:
     TypeError: Cannot read property 'toJSON' of undefined
```

Reply:
I think the failing tests are all ones that evaluate the outcome of `setTutorials()` or `showTutorialInModal()`, because those calls aren't waiting for done. 🤔 

I also think this is worrying me because the code change is harmless (open http://localhost:8080/?project=1233 in a browser and it works just fine) and working code shouldn't break the tests.

Reply:
Hang on, all the failing tests run when `rootStore.workflows.setActive()` resolves. Am I right in thinking that they should be waiting for `rootStore.tutorials.fetchTutorials()` to resolve, which happens slightly later? That might be the problem.

Reply:
I was going to say something like that! I've just picked it up again (sorry, am still feeling pretty rough), and looking at `it('should set the tutorial if there is a workflow')` - it's asserting that once a workflow is loaded, a tutorial _is also_ loaded, even though fetching a tutorial is happening in response to workflow loading

Reply:
I think those calls should be `rootStore.workflows.setActive(workflow).then(rootStore.tutorials.fetchTutorials).then…`.

Reply:
the other approach would be to create an observer on `tutorials.active`, look for an update where the tutorial id matches the expected one, and then call `done`; which might be preferable since you're not testing the exact implementation, just that a change happens and the expected thing happens in response

Reply:
No worries! Thanks very much for taking a look at this. It's probably a common case in the classifier tests. A promise resolves, triggering an observer that runs another promise and we don't want to test until the second promise has resolved.

Hope you feel better soon!

Reply:
Ha! Making my change doesn't fix all the tests, because spies are reset asynchronously, so tests for `calledOnce` start to break.

Reply:
My dumb solution, in the end, was to stub `fetchTutorials`, run `workflow.setActive` with the stub, then restore it and run `fetchTutorials` by hand. That stops the tests from running until after we've fetched tutorials, and preserves the spy history.